exercises/book/A3-traits-generics/mod.md

```diff
@@ -0,0 +1,151 @@
+
+**Questions**
+1. When is such a data structure more efficient than a standard `Vec`?
+2. What are the downsides, compared to `Vec`?
+
+Open the `exercises/A3/2-local-storage-vec` crate. It contains a `src/lib.rs` file, meaning this crate is a library. `lib.rs` contains a number of tests, which can be run by calling `cargo test`. Don't worry if they don't pass or even compile right now: it's your job to fix that in this exercise. Most of the tests are commented out right now, to enable a step-by-step approach. **Before you begin, have a look at the code and the comments in there, they contain various helpful clues.**
+
+### A3.2.A Defining the type
+Currently, the `LocalStorageVec` `enum` is incomplete. Give it two variants: `Stack` and `Heap`. `Stack` contains two named fields, `buf` and `usize`. `buf` will be the array containing `N` items of type `T`. The `Heap` variant has an unnamed field containing a `Vec<T>`. If you've defined the `LocalStorageVec` variants correctly, running `cargo test` should output something like
+
+```txt
+running 1 test
+test test::it_compiles ... ignored, This test is just to validate the definition of `LocalStorageVec`. If it compiles, all is OK
+
+test result: ok. 0 passed; 0 failed; 1 ignored; 0 measured; 0 filtered out; finished in 0.00s
+```
+
+This test does (and should) not run, but is just there for checking your variant definition.
+
+
+### A4.2.B `impl`-ing `From<Vec<T>`
+
+Uncomment the test `it_from_vecs`, and add an implementation for `From<Vec<T>>` to `LocalStorageVec<T>`. To do so, copy the following code in your `lib.rs` file and replace the `todo!` macro invocation with your code that creates a heap-based `LocalStorageVec` containing the passed `Vec<T>`.
+
+```rust
+impl<T, const N: usize> From<Vec<T>> for LocalStorageVec<T, N> {
+    fn from(v: Vec<T>) -> Self {
+        todo!("Implement me");
+    }
+}
+```
+
+**Question** 
+1. How would you pronounce the first line of the code you just copied in English?*
+
+Run `cargo test` to validate your implementation.
+
+### A3.2.C `impl LocalStorageVec`
+To make the `LocalStorageVec` more useful, we'll add more methods to it. Create an `impl`-block for `LocalStorageVec`. Don't forget to declare and provide the generic paramereters. For now, to make implementations easier, we will add a bound `T`, requiring that it implements `Copy` and `Default`. First off, uncomment the test called `it_constructs`. Make it compile and pass by creating a assosiated function called `new` on `LocalStorageVec` that creates a new, empty `LocalStorageVec` instance without heap allocation.
+
```

Comment:
I found it very useful to re-use `from` to implement `new`, you could add a hint about that.

Reply:
Let's have studens figure that out themselves. If they're stuck, they could always ask questions